processor/src/debug.rs

```diff
@@ -0,0 +1,47 @@
+use crate::{Felt, Process};
+use vm_core::STACK_TOP_SIZE;
+
+#[derive(Debug)]
+/// VmState holds a current process state information at a specific clock cycle.
+pub struct VmState {
+    pub stack: [Felt; STACK_TOP_SIZE],
+    pub clk: usize,
+    pub fmp: Felt,
+}
+
+/// Iterator that iterates through vm state at each step of the execution.
+/// This allows debuging or replaying ability to view various process state
+/// at each clock cycle.
+pub struct VmStateIterator {
+    process: Process,
+    current_clk: usize,
+}
+
+impl VmStateIterator {
+    pub(super) fn new(process: Process) -> Self {
+        Self {
+            process,
+            current_clk: 0,
+        }
+    }
+}
+
+impl Iterator for VmStateIterator {
+    type Item = VmState;
```

Comment:
I wonder if the iterator should return a result (i.g., `Result<VmState,ExecutionError>`). This way we can iterate over states until we get to an error.

Reply:
Since we're replaying after we execute, I'm not sure what is a way for us to be able to get that info?
Does the error that's returned always contain the clock cycle information, and so I can stop and return the same error?

Reply:
I think you should be able to get the last executed cycle from the `Process` struct even after the error.